File: 0203_Remove Linked List Elements/Solution.py
class ListNode:
    def __init__(self, val=0, next=None):
        self.val = val
        self.next = next

    @staticmethod
    def deserialize(data):
        """Convert a list into a linked list."""
        if not data:
            return None
        dummy = ListNode(0)
        current = dummy
        for val in data:
            current.next = ListNode(val)
            current = current.next
        return dummy.next

    @staticmethod
    def serialize(head):
        """Convert a linked list into a list."""
        result = []
        current = head
        while current:
            result.append(current.val)
            current = current.next
        return result


class Solution:
    def removeElements(self, head, val):
        """Remove all elements with the given value from the linked list."""
        # Create a dummy node to handle edge cases
        dummy = ListNode(0)
        dummy.next = head

        # Traverse the linked list
        current = dummy
        while current and current.next:
            if current.next.val == val:
                current.next = current.next.next  # Remove the node
            else:
                current = current.next
        return dummy.next


# Test cases
if __name__ == "__main__":
    solution = Solution()

    # Example 1: head = [1,2,6,3,4,5,6], val = 6
    head = ListNode.deserialize([1, 2, 6, 3, 4, 5, 6])
    print("Original List:", ListNode.serialize(head))
    head = solution.removeElements(head, 6)
    print("Modified List:", ListNode.serialize(head))

    # Example 2: head = [], val = 1
    head = ListNode.deserialize([])
    print("Original List:", ListNode.serialize(head))
    head = solution.removeElements(head, 1)
    print("Modified List:", ListNode.serialize(head))

    # Example 3: head = [7,7,7,7], val = 7
    head = ListNode.deserialize([7, 7, 7, 7])
    print("Original List:", ListNode.serialize(head))
    head = solution.removeElements(head, 7)
    print("Modified List:", ListNode.serialize(head))

File: 0203_Remove Linked List Elements/Solution.java
// package 0203_Remove Linked List Elements;

class ListNode {
    int val;
    ListNode next;

    ListNode() {}
    ListNode(int val) { this.val = val; }
    ListNode(int val, ListNode next) { this.val = val; this.next = next; }

    // Deserialize method to create a ListNode from a string
    public static ListNode deserialize(String data) {
        if (data.equals("[]")) {
            return null;
        }
        data = data.substring(1, data.length() - 1); // Remove brackets
        String[] nodes = data.split(",");
        ListNode dummy = new ListNode(0);
        ListNode current = dummy;
        for (String node : nodes) {
            current.next = new ListNode(Integer.parseInt(node.trim()));
            current = current.next;
        }
        return dummy.next;
    }

    // Serialize method to convert a ListNode to a string
    public static String serialize(ListNode head) {
        StringBuilder sb = new StringBuilder();
        sb.append("[");
        ListNode current = head;
        while (current != null) {
            sb.append(current.val);
            if (current.next != null) {
                sb.append(",");
            }
            current = current.next;
        }
        sb.append("]");
        return sb.toString();
    }
}

public class Solution {
    public ListNode removeElements(ListNode head, int val) {
        // Create a dummy node to simplify edge cases
        ListNode dummy = new ListNode(0);
        dummy.next = head;

        // Traverse the list
        ListNode current = dummy;
        while (current.next != null) {
            if (current.next.val == val) {
                // Skip the node with the target value
                current.next = current.next.next;
            } else {
                // Move to the next node
                current = current.next;
            }
        }
        return dummy.next; // Return the new head
    }

    public static void main(String[] args) {
        Solution solution = new Solution();

        // Example 1: head = [1,2,6,3,4,5,6], val = 6
        ListNode head = ListNode.deserialize("[1,2,6,3,4,5,6]");
        System.out.println("Original List: " + ListNode.serialize(head));
        head = solution.removeElements(head, 6);
        System.out.println("Modified List: " + ListNode.serialize(head));

        // Example 2: head = [], val = 1
        head = ListNode.deserialize("[]");
        System.out.println("Original List: " + ListNode.serialize(head));
        head = solution.removeElements(head, 1);
        System.out.println("Modified List: " + ListNode.serialize(head));

        // Example 3: head = [7,7,7,7], val = 7
        head = ListNode.deserialize("[7,7,7,7]");
        System.out.println("Original List: " + ListNode.serialize(head));
        head = solution.removeElements(head, 7);
        System.out.println("Modified List: " + ListNode.serialize(head));
    }
}

